# Nonparametric Methods

<!--

## Outline

- Introduction

- Distribution Estimation

  - The empirical distribution
    - As the nonparametric MLE
  - The Dirichlet Process
    - As a prior on random effects
    - Adaptive shrinkage

- Nonparametric Regression

  - Irregular problems and  bias-variance tradeoff: MLE is Useless
  - Curse of Dimensionality
  - Basis function expansions
    - Polynomial expansions
    - Natural cubic splines
    - Wavelets
    - Bayesian adaptive basis function expansions and BART
  - Smoothing Splines
  - Kernel Smoothing
    - Local linear smoothing
  - Gaussian process regression
    - Computation
  - Classification

- Density Estimation

  - Histograms
    - The Bayesian histogram
  - Kernel density estimators: why they work
  - Dirichlet process mixture models
  - Undersmoothing

-->

## Overview

In this set of notes, we will shift attention to various models that are referred to (broadly) as _nonparametric_. Speaking roughly, a model $\mathcal F = \{f_\theta : \theta \in \Theta\}$ is "nonparametric" if an $f \in \mathcal F$ "could be anything." Of course, this is a vague statement, and in reality the meaning of what a "nonparametric method" or a "nonparametric model" is varies according to context and the whims of whoever is writing about the topic. A couple of examples include:

1. Suppose $Z_1, \ldots, Z_n \iid F_0$ where $F_0 \in \mathcal F$ and $\mathcal F$ is the set of all possible distributions on $\mathbb R$, and we want to estimate the distribution of $Z_i$. This is the problem of _nonparametric distribution estimation_.

1. Ditto the above problem, except now we want to estimate the density $f_0$ of $F_0$. This is a problem of _nonparametric density estimation_, with the class $\mathcal F$ consisting of all distributions on $\mathbb R$ that admit a density.

1. Suppose we have
$$
  Y_i = \mu(X_i) + \epsilon_i
$$
where $\epsilon_i \iid \Normal(0, \sigma^2)$. We assume $\mu(x)$ is a function from $\mathbb R$ to $\mathbb R$. While this is probably more accurately referred to as _Gaussian semiparametric regression_, because $\mu(\cdot)$ is unrestricted some folks still think of this as a nonparametric problem.

Nonparametric models can be contrasted with _parametric_ models (which you should already be familiar) and _semiparametric_ models. Semiparametric models are sort of halfway between parametric and nonparametric models: they have a _parametric component_ indexed by a finite-dimensional parameter $\theta$, and a _nonparametric component_ indexed by an infinite-dimensional parameter $\eta$, which is often viewed as a nuisance parameter.

## Linear Smoothing, Basis Function Expansions, and Gaussian Processes

Much of the material in this section is borrowed from the notes of James Scott (with some minor changes on my part); the original is available [at this repository](https://github.com/jgscott/SDS383D/blob/master/exercises/exercises06-SDS383D.pdf). All errors and typos past this point are, of course, my own `r emo::ji("smile")`.

### The Bias-Variance Tradeoff

Let $\widehat \mu(x)$ be a noisy estimate of some function $\mu(x)$ at some point $x$ and definet he _mean-squared error_ of the estimate as
$$
  \MSE(\widehat\mu, \mu)
  =
  \E_\theta\left[ \{\widehat\mu(x) - \mu(x)\}^2 \right].
$$

:::{.exercise}

Show that $\MSE(\widehat\mu, \mu)$ decomposes into a _bias_ term and a _variance_ term $\MSE = B^2 + V$:
$$
  B = \E_\theta\{\widehat\mu(x)\} - \mu(x)
  \qquad \text{and} \qquad
  V = \Var\{\widehat\mu(x)\}.
$$

:::

This decomposition is referred to as the _bias-variance decomposition_. Generally speaking, nonparametric function estimation involves balancing these quantities: how small can we make the bias while keeping the variance manageable?

:::{.exercise name="Density Estimation With Histograms"}

Some people refer to the decomposition above as the _bias-variance tradeoff_. Why is this a tradeoff? Here's a simple example to convey the intuition.

Suppose we observe $Z_1, \ldots, Z_N$ from some distribution $F$ and want to estimate $f(0)$, the value of the probability density at $0$. Let $h$ be a small positive number, called the _bandwidth_, and define the quantity
$$
  \pi_h 
  = \Pr\left( -\frac{h}{2} < Z < \frac{h}{2}\right)
  = \int_{-\frac{h}{2}}^{\frac{h}{2}} f(z) \ dz.
$$
For small $h$ we have $\pi_h \approx h \, f(0)$ provided that $f(x)$ is continuous at $0$.

a. Let $M$ be the number of observations in a sample of size $N$ that fall within the interval $(-h/2, h/2)$. What is the distribution of $M$. What are its mean and variance in terms of $N$ and $\pi_h$? Propose a simple estimator $\widehat f(0)$ of $f(0)$ based on $M$.

b. Suppose we expand $f(z)$ in a second-order Taylor series about $0$:
$$
  f(z) = f(0) + f'(0) z + \frac{1}{2} f''(0) z^2 + O(z^3).
$$
Use this, together with the bias-variance decomposition, to show that
$$
  \MSE\{\widehat f(0), f(0)\}
  \approx
  A h^4 + \frac{B}{Nh}
$$
for constants $A$ and $B$ that you should (approximately) specify. What happens to the bias and variance when you make $h$ small? When you make $h$ big?

c. Use this result to derive an expression for the bandwidth that minimizes the mean-squared error as a function of $N$. You can approximate any constantst hat appear, but make sure you get the right functional dependence on the sample size.

:::

### Curve Fitting by Linear Smoothing

:::{.exercise}

Consider a nonlinear regression problem with one predictor and one response: $Y_i = \mu(X_i) + \epsilon_i$ where the $\epsilon_i$'s are mean-zero random variables.

a. Suppose you want to estimate the value of the regression function $\mu(x^\star)$ at some new point $x^\star$. Assume for the moment that $\mu(x)$ is linear and that both $Y_i$ and $X_i$ are mean $0$, in which case $Y_i = \beta \, X_i + \epsilon_i$.

    Recall that least-squares estimator for multiple regression. Show that for the one-predictor case, your prediction $\mu(x^\star) = \widehat \beta \, x^\star$ can be expressed as a _linear smoother_ of the form
$$
  \widehat \mu(x^\star)
  =
  \sum_{i=1}^N w(X_i, x^\star) \, Y_i
$$
    for any $x^\star$. Inspect the weighting function you derived. Briefly describe your understanding of how the reslting smoother behaves, compared with the smoother that arises from an alternate form of the weight function $w(X_i, x^\star)$:
$$
  w_K(X_i, x^\star) = 
  \begin{cases}
    1 / K \qquad & 
      \text{if $X_i$ is one of the $K$ closest sample points to $x^\star$},  \\
    0 \qquad & \text{otherwise}.
  \end{cases}
$$
    This is referred to as the _K-nearest neighbor_ smoother.

b. A _kernel function_ $K(x)$ is a smooth function satisfying
$$
  \int_{-\infty}^\infty K(x) \ dx = 1, 
  \qquad
  \int_{-\infty}^\infty x \, K(x) \ dx = 0.
  \qquad
  \int_{-\infty}^\infty x^2 \, K(x) \ dx < \infty.
$$
    A very simple example is the uniform kernel,
$$
  K(x) = \frac{1}{2} \, I(x) \qquad \text{where} \qquad
  I(x) = \begin{cases}
    1 \qquad & \text{if $|x| \leq 1$}, \\
    0 \qquad & \text{otherwise}.
  \end{cases}
$$
    Another common choice is the _Gaussian kernel_,
$$
  K(x) = \frac{1}{\sqrt{2 \pi}} \, \exp\left(-\frac{x^2}{2}\right).
$$
    Kernels are used as weighting functions for taking local averages; specifically, define the weighting function
$$
  w(X_i, x^\star) = \frac{1}{h}K\left(\frac{X_i - x^\star}{h}\right),
$$
    where $h$ is the bandwidth. Using this weighting function in a linear smoother is called _kernel regression_. (The weighting function gives unnormalized weights; you should also normalize the weights so that they sum to $1$.)

    __Write your own `R` function that will fit a kernel smoother for an arbitrary set of $(X_i, Y_i)$ pairs and arbitrary choice of (positive real) bandwidth $h$.__ You may choose your own kernel. Set up an `R` script that will simulate noisy data from some nonlinear function $Y_i = \mu(X_i) + \epsilon_i$; subtract the sample means from the simulated data; and use your function to fit the kernel smoother for some choice of $h$. Plot the estimated functions for a range of bandwidths wide enough to yield noticable differences in the qualitative behavior of the prediction functions.

:::

### Cross Validation

:::{.exercise}

Left unanswered so far in our previous study of kernel regression is the question: how does one choose the bandwidth $h$ used for the kernel?  Assume for now that the goal is to predict well, not necessarily to recover the truth.  (These are related but distinct goals.)  

a. Presumably a good choice of $h$ would be one that led to smaller predictive errors on fresh data.  Write a function or script that will: (1) accept an old ("training") data set and a new ("testing") data set as inputs; (2) fit the kernel-regression estimator to the training data for specified choices of $h$; and (3) return the estimated functions and the realized prediction error on the testing data for each value of $h$.  This should involve a fairly straightforward "wrapper" of the function you've already written.

b. Imagine a conceptual two-by-two table for the unknown, true state of affairs.  The rows of the table are "wiggly function" and "smooth function," and the columns are "highly noisy observations" and "not so noisy observations."  Simulate one data set (say, 500 points) for each of the four cells of this table, where the $X_i$'s take values in the unit interval.  Then split each data set into training and testing subsets.  You choose the functions.^[Trigonometric functions, for example, can be pretty wiggly if you make the period small.] Apply your method to each case, using the testing data to select a bandwidth parameter.  Choose the estimate that minimizes the average squared error in prediction, which estimates the mean-squared error:
$$
L_N(\widehat{\mu}) = \frac{1}{N^\star}\sum_{i=1}^{N^{\star}} (Y^{\star}_i - \widehat{Y}_i^{\star} )^2 \, ,
$$
where $(Y_i^{\star}, X_i^{\star})$ are the points in the test set, and $\widehat{Y}_i^{\star}$ is your predicted value arising from the model you fit using only the training data.  Does your out-of-sample predictive validation method lead to reasonable choices of $h$ for each case?

c. __Optional: ("leave-one-out lemma)"__ Splitting a data set into two chunks to choose $h$ by out-of-sample validation has some drawbacks.  (See if you can think of two; one is obvious, one is more subtle.)  Then consider an alternative: leave-one-out cross validation.  Define
$$
\mbox{LOOCV} = \sum_{i=1}^N \left( Y_i - \widehat{Y}_{i}^{(-i)} \right)^2 \, ,
$$
where $\widehat{Y}_{i}^{(-i)}$ is the predicted value of $Y_i$ obtained by omitting the $i^{\text{th}}$ pair $(X_i, Y_i)$ and fitting the model to the training data set of size $N - 1$.^[The intuition here is straightforward: for each possible choice of $h$, you have to predict each data point using all the others.  The bandwidth that with the lowest prediction error is the "best" choice by the LOOCV criterion.] This $\widehat{Y}_{i}^{(-i)}$ is contingent upon a particular bandwidth, and is obviously a function of $X_i$, but these dependencies are suppressed for ease of notation. This error metric looks expensive to compute: for each value of $h$, and for each data point to be held out, it seems as though you must fit a whole nonlinear regression model. But happily, there's a short-cut!

    Observe that for a linear smoother, we can write the whole vector of fitted values as $\widehat{Y} = H Y$, where $H$ is called the smoothing matrix (or "hat matrix") and $Y$ is the vector of observed outcomes. Deduce that, for any linear smoother,
$$
\mbox{LOOCV} = \sum_{i=1}^N \left( \frac{Y_i - \widehat{Y}_{i} } {1-H_{ii}} \right)^2 \, .
$$

    Hint: observe that for any matrix A whose rows\footnote{Remember that by default all vectors are column vectors. So when we write $a_i$, we mean the $i$th row of $A$ \emph{considered as a column vector}.  So as a row vector, it's $a_i^\top$.} are $a_i^\top, i = 1, \ldots N$, we have
$$
A^\top A = \sum_{i=1}^N a_i a_i^\top
$$
    Therefore, if we let $\bX_{(-i)}$ be the original feature matrix with the $i$th row $X_i^\top$ removed, we have
$$
\bX_{(-i)}^\top \bX_{(-i)} = \sum_{\ell \neq i} X_i X_i^\top = \sum_{\ell=1}^N X_\ell X_\ell^\top  - X_i X_i^\top = \bX^\top \bX - X_i X_i^\top
$$
    This hint is useful because computing $\widehat{Y}_{i}^{(-i)}$ requires inverting $\bX_{(-i)}^\top \bX_{(-i)}$.  This result, combined with the [Sherman-Morrison formula](https://en.wikipedia.org/wiki/Sherman%E2%80%93Morrison_formula), provides a path to do so, and ultimately to prove the leave-one-out lemma.  

    Like I said, proving this lemma is optional. But consider it a nice linear algebra exercise to try to prove it yourself.  However, you should use it for the following part...

d. Use the leave-one-out lemma to revisit the examples you simulated in Part B, using leave-one-out cross validation to select $h$ in each case.  Because of the leave-one-out lemma, you won't need to actually refit the model N times!

:::

### Local Polynomial Regression

Kernel regression has a nice interpretation as a "locally constant" estimator, obtained from locally weighted least squares.  To see this, suppose we observe pairs $(X_i, Y_i)$ for $i = 1, \ldots, N$ from our new favorite model, $Y_i = \mu(X_i) + \epsilon_i$ and wish to estimate the value of the underlying function $\mu(x)$ at some point $x$ by weighted least squares.  Our estimate is the scalar^[Because we are only talking about the value of the function at a specific point $x$, not the whole function.] quantity
$$
  \widehat\mu(x) = a = 
  \arg \min_{\mathbb R} \sum_{i=1}^N w_i (Y_i - a)^2 \, ,
$$
where the $w_i$ are the normalized weights (i.e.~they have been rescaled to sum to 1 for fixed $x$).  Clearly if $w_i = 1/N$, the estimate is simply $\bar{Y}$, the sample mean, which is the "best" globally constant estimator.  Using elementary calculus, it is easy to see that if the unnormalized weights are
$$
  w_i \equiv w(x, X_i) = \frac{1}{h} K \left( \frac{X_i - x}{h} \right)  \, ,
$$
then the solution is exactly the kernel-regression estimator.

:::{.exercise}

a. A natural generalization of the locally constant regression is local polynomial regression. For points $u$ in a neighborhood of the target point $x$, define the polynomial
$$
  g_{x}(u; a) = a_0 + \sum_{k=1}^D a_k \, (u-x)^k
$$
for some vector of coefficients $a = (a_0, \ldots, a_D)$.  As above, we will estimate the coefficients $a$ in $g_{x}(u; a)$ at some target point $x$ using weighted least squares:
$$
  \widehat{a} = 
    \arg \min_{\mathbb R^{D+1}} 
    \sum_{i=1}^N w_i \left\{ Y_i - g_{x}(X_i; a)  \right\}^2 \, ,
$$
where $w_i \equiv w(X_i, x)$ are the kernel weights defined just above, normalized to sum to one.^[We are fitting a different polynomial function for every possible choice of $x$.  Thus $\widehat{a}$ depends on the target point $x$, but we have suppressed this dependence for notational ease.] Derive a concise (matrix) form of the weight vector $\widehat{a}$, and by extension, the local function estimate $\widehat{\mu}(x)$ at the target value $x$.^[Observe that at the target point $x$, $g_x(u = x; a) = a_0$.  That is, only the constant term appears.  But this is not the same thing as fitting only a constant term!] Life will be easier if you define the matrix $R_x$ whose $(i,j)^{\text{th}}$ entry is $(X_i - x)^j$, and remember that (weighted) polynomial regression is the same thing as (weighted) linear regression with a polynomial basis.

    Note: If you get bogged down doing the general polynomial case, just try the linear case.

b. From this, conclude that for the special case of a local linear estimator ($D = 1$), we can write $\widehat\mu(x)$ as a linear smoother of the form
$$
  \widehat\mu(x) 
  = \frac{\sum_{i=1}^N w_i(x) \, Y_i}
         {\sum_{i = 1}^N w_i(x)},
$$
where the unnormalized weights are
$$
\begin{aligned}
  w_i(x) &= K \left( \frac{x-X_i}{h} \right) \left\{  s_2(x) - (X_i-x) s_1(x) \right\}\\
  s_j(x) &= \sum_{i=1}^N K \left( \frac{x-X_i}{h} \right) (X_i-x)^j \, .
\end{aligned}
$$

c. Suppose that the residuals have constant variance $\sigma^2$ (that is, the spread of the residuals does not depend on $x$).  Derive the mean and variance of the sampling distribution for the local polynomial estimate $\widehat{\mu}(x)$ at some arbitrary point $x$.  Note: the random variable $\widehat{\mu}(x)$ is just a scalar quantity at $x$, not the whole function.

d. We don't know the residual variance, but we can estimate it.  A basic fact is that if $X$ is a random vector with mean $\mu$ and covariance matrix $\Sigma$, then for any symmetric matrix $Q$ of appropriate dimension, the quadratic form $X^\top Q X$ has expectation
$$
E(X^\top Q X) = \mbox{tr}(Q \Sigma) + \mu^\top Q \mu \, .
$$
Write the vector of residuals as $R = Y - \widehat{Y} = Y - HY$, where $H$ is the smoothing matrix.  Compute the expected value of the estimator
$$
\widehat{\sigma}^2 = \frac{\|r\|^2}{n - 2\mbox{tr}(H) + \mbox{tr}(H^\top H)} \, ,
$$
and simplify things as much as possible.  Roughly under what circumstances will this estimator be nearly unbiased for large $n$?  Note: the quantity $2\mbox{tr}(H) - \mbox{tr}(H^\top H)$ is often referred to as the "effective degrees of freedom'" in such problems.

e. Write code that fits the local linear estimator using a Gaussian kernel for a specified choice of bandwidth $h$. Then load the data in ``utilities.csv'' into `R`.^[On the class GitHub site.] This data set shows the monthly gas bill (in dollars) for a single-family home in Minnesota, along with the average temperature in that month (in degrees F), and the number of billing days in that month.  Let $Y_i$ be the average daily gas bill in a given month (i.e.~dollars divided by billing days), and let $X_i$ be the average temperature. Fit $Y_i$ versus $X_i$ using local linear regression and some choice of kernel.  Choose a bandwidth by leave-one-out cross-validation.

f. Inspect the residuals from the model you just fit.  Does the assumption of constant variance (homoskedasticity) look reasonable?  If not, do you have any suggestion for fixing it?

g. Put everything together to construct an approximate point-wise 95\% confidence interval for the local linear model (using your chosen bandwidth) for the value of the function at each of the observed points $X_i$ for the utilities data.  Plot these confidence bands, along with the estimated function, on top of a scatter plot of the data.^[It's fine to use Gaussian critical values for your confidence set.]

:::

## Nonparametric Bayesian Smoothing Via Gaussian Processes

An alternative setup to the smoothing problem is to instead think about the parameter space for $\mu(\cdot)$ as a _function space_ $\mu \in \mathscr F$ on a subspace $\mathcal X$ of $\Reals^P$; examples of function spaces include

- The space of continuous functions on a space $\mathcal X$.
- The space of functions on $\mathcal X$ that are twice-continuously differentiable.
- The space of square-integrable functions on $\mathcal X$ (i.e., the set of $\mu$'s such that $\int \mu(x)^2 \, f(x) \ dx$ for some density $f$ on $\mathcal X$).

A criteria for a prior to be "flexible" from this perspective is that the support of the prior $\Pi(d\mu)$ should have _large support_ in the sense that an arbitrary $\mu_0$ can be well-approximated by a sample from the prior; for example, we might ask that
$$
  \Pi(\sup_{x \in \mathcal X} |\mu_0(x) - \mu(x)| < \epsilon) > 0
$$
for all $\mu_0 \in \mathscr F$ and all $\epsilon > 0$. A common choice of prior $\Pi(d\mu)$ that can accomplish these goals is a Gaussian process prior.

<!--

A _Gaussian process_ is a collection of random variables $\{\mu(x) : x \in \mathcal X\}$ such that, for any finite collection of indices $x_1, \ldots, x_N \in \mathcal X$, the random vector $(\mu(x_1), \ldots, \mu(x_N))^\top$ has a multivariate normal distribution. It is a generalization of the multivariate normal distribution to infinite-dimensional spaces. The set $\mathcal X$ is called the _index set_ or _state space_ of the process, and need not be countable.

-->

:::{.definition}

Let $m : \mathcal X \to \mathbb R$ and $K: \mathcal X^2 \to \mathbb R$.

A random function $\mu : \mathcal X \to \mathbb R$ is said to be a _Gaussian process_ if, for any _finite_ set $D = \{x_1, \ldots x_D\}$ we have
$$
  \mu(\mathbf x) =
  \Normal\{m(\mathbf x), K(\mathbf x, \mathbf x)\}
$$
where $\mathbf x = (x_1, \ldots, x_D)^\top$, $\mu(\mathbf x) = (\mu(x_1), \ldots, \mu(x_D))^\top$, $m(\mathbf x) = (m(x_1), \ldots, m(x_D))^\top$, and $K(\mathbf x, \mathbf x')$ is a covariance matrix with $(i,j)^{\text{th}}$ entry $K(x_i, x'_j)$. The function $K(\cdot, \cdot)$ is referred to as a _covariance function_.

To denote this fact, we write $\mu \sim \GP(m, K)$.

:::

(Note: not all functions $K(x,x')$ can be used for this purpose, as the definition requires that $K(\mathbf x, \mathbf x')$ always be a valid covariance matrix. The covariance function is also sometimes referred to as a _kernel_ function.)

:::{.exercise}

Suppose that $Y_i \stackrel{\text{indep}}{\sim} \Normal\{\mu(X_i), \sigma^2\}$ conditional on $\bX$ for $i = 1, \ldots, N$ and $\mu$. Let $\bX = (X_1, \ldots, X_N)$ and $\bY = (Y_1, \ldots, Y_N)$.

Show that the posterior distribution of $\mu$ is given by
$$
  [\mu \mid \bX, \bY, \sigma^2]
  \sim
  \GP(m^\star, K^\star)
$$
where
$$
\begin{aligned}
  m^\star(x) &=
    m(x) +
      K(x, \bX) \{K(\bX, \bX) +
      \sigma^2 \Identity\}^{-1} \{\mathbf Y - m(\mathbf X)\} \\
  K^\star(x, x') &= K(x, x') - [[XXX]]
\end{aligned}
$$

:::

:::{.exercise}

Read up on the [Matérn class](https://en.wikipedia.org/wiki/Mat%C3%A9rn_covariance_function) of covariance functions. The Matérn class has the _squared exponential_ covariance function as a special case:
$$
  K(x, x') = \sigma_\mu^2 \exp \left\{ -\frac{1}{2} \sum_{j=1}^P \left(\frac{x_j - x'_j}{h}\right)^2 \right\} + \sigma_\delta^2 \,  \delta(x, x') \, ,
$$

:::

## Basis Function Expansions

<!-- 

\begin{enumerate}[(A)]
\item   Read up on the \href{https://en.wikipedia.org/wiki/Matrn_covariance_function}{Matern class}\footnote{\url{https://en.wikipedia.org/wiki/Matern_covariance_function}} of covariance functions.  The Matern class has the \textit{squared exponential} covariance function as a special case:
$$
C_{SE}(x_1, x_2) = \tau_1^2 \exp \left\{ - \frac{1}{2} \left( \frac{d(x_1, x_2)}{b} \right)^2 \right\} + \tau^2_2 \delta(x_1, x_2) \, ,
$$
where $d(x_1, x_2) = \Vert x_1 - x_2 \Vert_2$ is Euclidean distance (or just $|x-y|$ for scalars).  The constants $(b, \tau^2_1, \tau^2_2)$ are often called \textit{hyperparameters}, and $\delta(a,b)$ is the Kronecker delta function that takes the value 1 if $a=b$, and 0 otherwise.  But usually this covariance function generates functions that are ``too smooth,'' and so we use other covariance functions in the Matern class as a default.\footnote{See the speed comparison in kernel-benchmark.R on the class GitHub site if you want to see how Rcpp can be used to speed things up here.  My code is for the squared-exponential covariance function.}

Let's start with the simple case where $\mathcal{X} = [0,1]$, the unit interval.  Write a function that simulates a mean-zero Gaussian process on $[0,1]$ under the squared exponential covariance function.  The function will accept as arguments: (1) finite set of points $x_1, \ldots, x_N$ on the unit interval; and (2) a triplet $(b, \tau^2_1, \tau^2_2)$. It will return the value of the random process at each point: $f(x_1), \ldots, f(x_N)$.

Use your function to simulate (and plot) Gaussian processes across a range of values for $b$, $\tau^2_1$, and $\tau^2_2$.  Try starting with a very small value of $\tau^2_2$ (say, $10^{-6}$) and playing around with the other two first.  On the basis of your experiments, describe the role of these three hyperparameters in controlling the overall behavior of the random functions that result.  What happens when you try $\tau^2_2 = 0$? Why?  If you can fix this, do---remember our earlier discussion on different ways to simulate the MVN.

Now simulating a few functions with a different covariance function, the Mat\'ern with parameter $5/2$:
$$
C_{M52}(x_1, x_2) = \tau_1^2 \left\{ 1 + \frac{\sqrt{5}d}{b} + \frac{5d^2}{3b^2} \right\} \exp\left( \frac{-\sqrt{5}d}{b} \right) + \tau^2_2 \delta(x_1, x_2) \, ,
$$
where $d = \Vert x_1 - x_2 \Vert_2$ is the distance between the two points $x_1$ and $x_2$.  Comment on the differences between the functions generated from the two covariance kernels.

\item Suppose you observe the value of a Gaussian process $f \sim \mbox{GP}(m,C)$ at points $x_1, \ldots, x_N$.  What is the conditional distribution of the value of the process at some new point $x^{\star}$?  For the sake of notational ease simply write the value of the $(i,j)$ element of the covariance matrix as $C_{i,j}$, rather than expanding it in terms of a specific covariance function.

\item Prove the following lemma.

\begin{lemma}
Suppose that the joint distribution of two vectors $y$ and $\theta$ has the following properties: (1) the conditional distribution for $y$ given $\theta$ is multivariate normal, $(y \mid \theta) \sim N(R\theta, \Sigma)$; and (2) the marginal distribution of $\theta$ is multivariate normal, $\theta \sim N(m,V)$.  Assume that $R$, $\Sigma$, $m$, and $V$ are all constants.  Then the joint distribution of $y$ and $\theta$ is multivariate normal.
\end{lemma}

\end{enumerate}


\section{In nonparametric regression and spatial smoothing}

\begin{enumerate}[(A)]

\item Suppose we observe data $y_i = f(x_i) + \epsilon_i$, $\epsilon_i \sim N(0, \sigma^2)$, for some unknown function $f$. Suppose that the prior distribution for the unknown function is a mean-zero Gaussian process: $f \sim \mbox{GP}(0, C)$ for some covariance function $C$.  Let $x_1, \ldots, x_N$ denote the previously observed $x$ points.  Derive the posterior distribution for the random vector $[f(x_1), \ldots, f(x_N)]^\top$, given the corresponding outcomes $y_1, \ldots, y_N$, assuming that you know $\sigma^2$.

\item As before, suppose we observe data $y_i = f(x_i) + \epsilon_i$, $\epsilon_i \sim N(0, \sigma^2)$, for $i=1, \ldots, N$.  Now we wish to predict the value of the function $f(x^{\star})$ at some new point $x^{\star}$ where we haven't seen previous data.  Suppose that $f$ has a mean-zero Gaussian process prior, $f \sim GP(0, C)$.  Show that the posterior mean $E\{ f(x^{\star}) \mid y_1, \ldots, y_N \}$ is a linear smoother, and derive expressions both for the smoothing weights and the posterior variance of $f(x^{\star})$.

\item Go back to the utilities data, and plot the pointwise posterior mean and 95\% posterior confidence interval for the value of the function at each of the observed points $x_i$ (again, superimposed on top of the scatter plot of the data itself).  Choose $\tau^2_2$ to be very small, say $10^{-6}$, and choose $(b, \tau^2_1)$ that give a sensible-looking answer.\footnote{If you're bored with the utilities data, instead try the data in ethanol.csv, in which the NOx emissions of an ethanol engine are measured as the engine's fuel-air equivalence ratio (E in the data set) is varied.  Your goal would be to model NOx as a function of E using a Gaussian process. }

\item Let $y_i = f(x_i) + \epsilon_i$, and suppose that $f$ has a Gaussian-process prior under the Matern(5/2) covariance function $C$ with scale $\tau^1_2$, range $b$, and nugget $\tau^2_2$.  Derive an expression for the marginal distribution of $y = (y_1 \ldots, y_N)$ in terms of $(\tau^2_1, b, \tau^2_2)$, integrating out the random function $f$.  This is called a marginal likelihood.

\item Return to the utilities or ethanol data sets. Fix $\tau^2_2 = 0$, and evaluate the log of the marginal likelihood function $p(y \mid \tau^2_1, b)$ over a discrete 2-d grid of points.\footnote{Don't just use a black-box optimizer; we want to make sure we get the best solution if there are multiple modes.} If you're getting errors in your code with $\tau^2_2 = 0$, use something very small instead.  Use this plot to choose a set of values $(\widehat{\tau^2_1}, \widehat{b})$ for the hyperparameters.  Then use these hyperparameters to compute the posterior mean for $f$, given $y$.  Comment on any lingering concerns you have with your fitted model.

\item In \verb|weather.csv| you will find data on two variables from 147 weather stations in the American Pacific northwest.
\begin{compactdesc}
\item[pressure]: the difference between the forecasted pressure and the actual pressure reading at that station (in Pascals)
\item[temperature]: the difference between the forecasted temperature and the actual temperature reading at that station (in Celsius)
\end{compactdesc}
There are also latitude and longitude coordinates of each  station.  Fit a Gaussian process model for each of the temperature and pressure variables.  Choose hyperparameters appropriately.  Visualize your fitted functions (both the posterior mean and posterior standard deviation) on a regular grid using something like a contour plot or color image.  Read up on the \verb|image|, \verb|filled.contour|, or \verb|contourplot|\footnote{in the lattice library} functions in R.  An important consideration: is Euclidean distance the appropriate measure to go into the covariance function?  Or do we need separate length scales for the two dimensions, i.e.
$$
d^2(x, z) = \frac{(x_1 - z_1)^2}{b_1^2} +  \frac{(x_2 - z_2)^2}{b_2^2} \, .
$$
Justify your reasoning for using Euclidean distance or this ``nonisotropic'' distance.


%\item In \verb|droslong.csv|, you will find a small subset of a time-course DNA microarray experiment.  The gene-expression profiles of 2000 different genes in the fruit fly (Drosophila) genome are tracked over time during embryogenesis; you are getting data on 14 of these genes, organized in three groups (think of these as marking which cellular pathway that gene influences).  For each gene at each time point, there are 3 ``technical replicates''---that is, three copies of the same biological material from the same fly, run through the same process to measure gene expression.
%
%The question of interest is: how does each gene's expression profile change over time, as the process of embryogenesis unfolds?  Propose a hierarchical model for this data that properly reflects its structure.
%
%A nice graphics package is the ``lattice'' library.  Install and load this; then try commands such as
%\begin{verbatim}
%xyplot(log2exp~time | gene, data=droslong)
%xyplot(log2exp~time | group, data=droslong)
%\end{verbatim}
%to begin exploring the structure of this data.

\end{enumerate}

-->

<!-- ## Bayesian Nonparametric Estimation of a Distribution -->

<!-- The most general nonparametric estimation problem is the problem of estimating an arbitrary _distribution_ $F$; [[ADD MOTIVATION FOR CONSIDERING THIS PROBLEM]]. -->

<!-- A straight-forward estimate of $F$ is the empirical distribution $\EmpF$, which you should be familiar with at this point, so we won't go into much detail here. Instead, we will focus primarily on addressing this problem within the Bayesian framework. The main tool here is the _Dirichlet process_ (and its variants). -->

<!-- :::{.definition} -->

<!-- __Dirichlet process:__ Let $\mathcal Z$ denote the sample space for some random variable $Z_1$, let $\alpha > 0$, and let $H$ be a probability distribution on $\mathcal Z$. We say that a _random probability measure_ (RPM) $F$ is a _Dirichlet process_ on $\mathcal Z$ if, for any _finite measurable partition_ $A_1, \ldots, A_k$ of $\mathcal Z$ we have -->
<!-- $$ -->
<!--   (F(A_1), \ldots, F(A_k)) -->
<!--   \sim -->
<!--   \Dirichlet\{\alpha \, H(A_1), \ldots, \alpha \, H(A_k)\}. -->
<!-- $$ -->

<!-- ::: -->

<!-- The definition [[XXX]] is conceptually pleasant but can sometimes be difficult to work with and masks some of the more important properties of the Dirichlet process. An equivalent definition is given below, which is "constructive" in nature. -->

<!-- :::{.definition} -->

<!-- __Alternate definition of the Dirichlet Process:__ Let $\mathcal Z$ denote the sample space for some random variable $Z_1$, let $\alpha > 0$, and let $H$ be a probability distribution on $\mathcal Z$. We say that an RPM $F$ is a _Dirichlet process_ on $\mathcal Z$ if we can write -->
<!-- $$ -->
<!--   F(dz) = \sum_{k = 1}^\infty \omega_k \, \delta_{\theta_k}(dz) -->
<!-- $$ -->
<!-- where $\theta_k \iid H$ and the $\omega_k$'s follow the _Sethuramann stick-breaking construction_ -->
<!-- $$ -->
<!--   \omega_k = \omega'_k \prod_{j < k} (1 - \omega'_j) -->
<!--   \qquad \text{where} \qquad -->
<!--   \omega'_k \sim \Beta(1, \alpha). -->
<!-- $$ -->
<!-- The prior for $\omega = (\omega_1, \omega_2, \ldots)$ is called the _[[XXX]] distribution_ and we write $\omega \sim \text{GEM}(\alpha)$ to denote this. -->

<!-- ::: -->

<!-- __Why is this true?__ A rigorous proof of the fact that the two definitions above are equivalent is beyond our scope, but if you are curious as to how one might guess that this is true note that under the original definition of the Dirichlet process it is possible to show that -->
<!-- $$ -->
<!--   F(dz) \stackrel{d}{=} \omega_1' \, \delta_{\theta} + (1 - \omega_1') \, F(dz). -->
<!-- $$ -->
<!-- where $\omega_k' \iid \Beta(1, \alpha)$ (it's not beyond your abilities to prove this as a homework exercise, but we have better uses of our time). We can then basically repeatedly apply this to write -->
<!-- $$ -->
<!--   F(dz) = \sum_{k=1}^K \omega_k \, \delta_{\theta_k} + \left(\prod_{j = 1}^K (1 - \omega'_k)\right) \, F(dz). -->
<!-- $$ -->
<!-- so that "letting $K \to \infty$" the first term goes to the Sethurmann stick-breaking construction, while the second term goes to $0$. This was basically Sethuramann's original reasoning, but it took a rather long time to publish the result, partially because getting all of the technical details correct is a bit tricky and, at the time, of unclear practical value. -->

<!-- ### The Posterior Distribution for the Dirichlet Process -->

<!-- Aside from its elegant definition, the reason BNP inference for a distribution revolves around the Dirichlet process is that it is _conjugate to iid sampling._ -->

<!-- :::{.exercise} -->

<!-- Suppose that $Z_1, \ldots, Z_N \iid F$ given $F$ and that $F \sim \DP(\alpha, H)$. -->

<!-- (a) Show that, for an arbitrary set $A$, we have -->
<!-- $$ -->
<!--   \E\{F(A)\} = H(A). -->
<!-- $$ -->
<!-- The distribution $H$ is referred to as the _prior mean_ of $F$; explain why this terminology makes sense. -->

<!-- (b) Show that, for an arbitrary set $A$, we have -->
<!-- $$ -->
<!--   \Var\{F(A)\} = \frac{H(A) \, \{1 - H(A)}}{\alpha + 1}. -->
<!-- $$ -->
<!-- The parameter $\alpha$ is referred as the _concentration_ or _precision_ parameter; explain why this terminology makes sense. -->

<!-- (a) Show that the posterior distribution of $[F \mid Z_1, \ldots, Z_N]$ is a Dirichlet process with updated paraemters -->
<!-- $$ -->
<!--   \alpha^\star = \alpha + N -->
<!--   \qquad \text{and} \qquad -->
<!--   H^\star = \frac{\alpha}{\alpha + N} \, H + \frac{N}{\alpha + N} \, \EmpF_N. -->
<!-- $$ -->

<!-- ::: -->

<!-- :::{.exercise} -->

<!-- Revist the problem [[XXX]] -->

<!-- ::: -->

<!-- ## Nonparametric Estimation of a Regression Function -->

<!-- $$ -->
<!--   Y_i = \mu(X_i) + \epsilon_i -->
<!-- $$ -->

<!-- ### The Curse of Dimensionality -->

<!-- The parametric and nonparametric esimation problems are quite different in practice; in particular, nonparametric estimation usually _much more difficult_ than parametric estimation. -->

<!-- One complication is the so-called _curse of dimensionality_. [[XXX]] -->

<!-- ### Basis Function Expansions -->

<!-- $$ -->
<!-- \begin{aligned} -->
<!--   Y_i = \beta_0 + \psi(X_i)^\top \beta + \epsilon_i, -->
<!--   \qquad \E(\epsilon_i \mid X_i) = 0 -->
<!-- \end{aligned} -->
<!-- $$ -->

<!-- ```{r, echo = FALSE, message=FALSE, fig.cap="A set of cubic spline basis functions."} -->
<!-- plot_spline_basis <- function(num_basis) { -->
<!--   require(splines) -->
<!--   require(latex2exp) -->
<!--   x <- seq(from = 0, to = 1, length = 500) -->
<!--   knots <- quantile(x, probs = seq(.1, .9, length.out = num_basis + 2)) -->
<!--   X <- ns(x, knots = knots, intercept = FALSE) -->

<!--   plot(x, X[,1], -->
<!--     type = 'l', -->
<!--     ylim = c(0, max(X)), -->
<!--     xlab = 'x', -->
<!--     ylab = TeX("$\\psi_j(x)$") -->
<!--   ) -->
<!--   for(i in 2:num_basis) { -->
<!--     lines(x, X[,i]) -->
<!--   } -->
<!-- } -->

<!-- plot_spline_basis(10) -->
<!-- ``` -->

<!-- :::{.exercise} -->

<!-- A particular type of basis function expansion takes $\psi(x) = (\psi_1(x), \ldots, \psi_L(x))$ to correspond to step functions: -->
<!-- $$ -->
<!--   \psi_\ell(x) -->
<!--   = -->
<!--   \begin{cases} -->
<!--     1 \qquad & \mbox{if } x \in A_\ell, \\ -->
<!--     0 \qquad & \mbox{otherwise}. -->
<!--   \end{cases} -->
<!-- $$ -->

<!-- ::: -->

<!-- ### Smoothing Splines -->

<!-- An alternate to the basis function approach to estimating $\mu(x)$ is to use -->

<!-- ### Kernel Smoothing -->

<!-- ### Bayesian Methods -->

<!-- For the most part, the methods above have Bayesian analogs, and these methods typically have similar operating characteristics as the Frequentist solutions. For example, we could [[XXX: ADD MATERIAL ON BASIS FUNCTION EXPANSION]] -->

<!-- [[MATERIAL ON GPs]] -->

<!-- :::{.exercise} -->

<!-- A common choice of covariance function for a Gaussian process is the _squared exponential kernel_ -->
<!-- $$ -->
  <!-- K(x, x') -->
  <!-- = -->
  <!-- \sigma^2_\mu \exp\left(-\rho \sum_{j=1}^P (x_j - x'_j)^2\right). -->
<!-- $$ -->

<!-- Code for generating some approximate samples from this prior distribution is given below. -->

<!-- [[ADD CODE HERE]] -->

<!-- a. Suppose that two points $x$ and $x'$ are identical except that $x_j \ne x'_j$ for some $j$. What is the correlation between $\mu(x)$ and $\mu(x')$ as a function of $(\sigma^2_\mu, \rho)$? -->

<!-- a. [[ADD A SECOND PART HERE]] -->

<!-- ::: -->

<!-- The main challenge with using Gaussian processes is that they require the computation of $\{K(\bX, \bX) + \sigma^2\, \Identity\}^{-1}(\bY - m(\bX))$. Ostensibly, this requires computing the inverse of an $N \times N$ matrix, which is a very expensive operation, having computational complexity that scales like $N^3$ and memory requirement that scales like $N^2$; so, for example, if you have $10,000$ observations, you expect to need $100,000,000$ units of storage available and to wait $1,000,000,000,000$ units of time to compute the posterior. Needless to say, this is very expensive compared to (say) computing the least squares estimator $\widehat \beta$, which can be done with constant storage and compute time that scales with $N$. -->

<!-- Because of this, there is a cottage industry focused no developing methods for fast inference with Gaussian processes - the two main strategies are to either (i) design $K(\mathbf x, \mathbf x')$ so that required computations can be performed quickly or (ii) approximate the inverse. -->



<!-- ## Nonparametric Estimation of a Density -->

<!-- ### Kernel Density Estimation -->

<!-- ### Mixture Models -->

<!-- ### Bayesian Methods -->

<!-- The Dirichlet process is unfortunately not adequate for the purpose of modeling a density because it assigns prior probability $1$ to discrete distributions. Instead, most Bayesian approaches to density estimation are based on mixture models like [[XXX]], with a Dirichlet process used to model the latent variables; that is, we set -->
<!-- $$ -->
<!-- \begin{aligned} -->
<!--   [Y_i \mid \bm\theta F] &\iid p_{\theta_i}(y), \\ -->
<!--   [\theta_i \mid F] &\iid G, \\ -->
<!--   G &\sim \DP(\alpha, H) -->
<!-- \end{aligned} -->
<!-- $$ -->
<!-- Or, equivalently -->
<!-- $$ -->
<!--   f_G(y) -->
<!--   = \int p_\theta(y) \ G(d\theta) -->
<!--   = \sum_{k = 1}^\infty \omega_k \, p_{\theta_k}(y). -->
<!-- $$ -->

<!-- [[MORE STUFF HERE]] -->

<!-- :::{.exercise} -->

<!-- Use the package `dirichletprocess` to fit a Dirichlet process mixture model to the `galaxy` dataset. Then, plot the following quantities together: -->

<!-- - The posterior mean of the density along a grid of 1000 points extending from _xxx_ to _xxx_. -->

<!-- - A 95% credible bands for the density using the same grid. -->

<!-- ::: -->


